### PR TITLE
Request object rewrite: RequestType enum

### PR DESCRIPTION
**Pull request by agrinman**
Use the RequestType enum to switch between different types of requests (i.e. ssh and git, me, unpair). 
Removes the optional [X]Request members from the Request object.

Additionally, implement error UI for NotifyUI in case of invalid requests.

### CODE REVIEW

File: Kryptonite/Silo.swift [hunk 1]
```diff
@@ -274,31 +282,33 @@ class Silo {
...
             
             var pgpPublicKey:Data?
             if let pgpUserID = meRequest.pgpUserId {
                 pgpPublicKey = try keyManager.loadPGPPublicKey(for: pgpUserID).packetData
             }
             
-            me = MeResponse(me: MeResponse.Me(email: try keyManager.getMe(), publicKeyWire: try keyManager.keyPair.publicKey.wireFormat(), pgpPublicKey: pgpPublicKey))
+            responseType = .me(MeResponse(me: MeResponse.Me(email: try keyManager.getMe(), publicKeyWire: try keyManager.keyPair.publicKey.wireFormat(), pgpPublicKey: pgpPublicKey)))
+
+        default:
```
Comment (kcking) [inline Kryptonite/Silo.swift]: are you sure you want `default` here as well?

File: Kryptonite/Silo.swift [hunk 2]
```diff
@@ -274,31 +282,33 @@ class Silo {
...
+            throw ResponseNotNeededError()
```
Comment (kcking) [inline Kryptonite/Silo.swift]: isn't this expected behavior? maybe we can return `Response?`

File: Notify/NotificationService.swift [hunk 3]
```diff
@@ -71,45 +72,34 @@ class NotificationService: UNNotificationServiceExtension {
                         self.bestAttemptMutex.lock {
                             
                             let content = UNMutableNotificationContent()
+                            let (noteSubtitle, noteBody) = unsealedRequest.notificationDetails()
                             
-                            var errorMessage:String?
+                            content.subtitle = noteSubtitle
                             
-                            // approved
+                            // auto-approved
```
Comment (kcking) [inline Notify/NotificationService.swift]: I wouldn't call this auto-approved -- isn't it just cached (but potentially rejected)?

File: NotifyUI/NotificationViewController.swift [hunk 4]
```diff
@@ -52,26 +52,25 @@ class NotificationViewController: UIViewController, UNNotificationContentExtensi
...
+                switch gitSignRequest.git {
                 case .commit(let commit):
                     showView(type: ContainerType.commit(commit), deviceName: sessionName)
                 case .tag(let tag):
                     showView(type: ContainerType.tag(tag), deviceName: sessionName)
                 }
-                
-            } else {
+            default:
                 throw InvalidNotificationError.requestTypeUnknown(sessionName)
```
Comment (kcking) [inline NotifyUI/NotificationViewController.swift]: will this happen for me/unpair requests?

File: Kryptonite/Request.swift [hunk 5]
```diff
@@ -15,80 +15,116 @@ struct Request:Jsonable {
     var unixSeconds:Int
     var sendACK:Bool
     var version:Version
-    var sign:SignRequest?
-    var gitSign:GitSignRequest?
-    var me:MeRequest?
-    var unpair:UnpairRequest?
+    var type:RequestType
```
Comment (agrinman) [inline Kryptonite/Request.swift]: 👍

File: Kryptonite/Silo.swift [hunk 6]
```diff
@@ -75,80 +75,85 @@ class Silo {
...
-        }
-
+        
+        // ensure request has not expired
         let now = Date().timeIntervalSince1970
         if abs(now - Double(request.unixSeconds)) > Properties.requestTimeTolerance {
             throw InvalidRequestTimeError()
         }
         
+        // check if the request has already been receieved and cached
```
Comment (agrinman) [inline Kryptonite/Silo.swift]: 👍

File: Kryptonite/Silo.swift [hunk 7]
```diff
@@ -75,80 +75,85 @@ class Silo {
...
             return
-        }
-
-        if request.isNoOp() {
+            
+        case .git where Policy.needsUserApproval(for: session):
+            try handleRequestRequiresApproval(request: request, session: session, communicationMedium: communicationMedium, completionHandler: completionHandler)
             return
+            
+        default:
```
Comment (agrinman) [inline Kryptonite/Silo.swift]: should not be default 👍

File: Kryptonite/Silo.swift [hunk 8]
```diff
@@ -75,80 +75,85 @@ class Silo {
...
             
-            if let error = response.gitSign?.error {
+            if let error = gitSign.error {
                 Policy.notifyUser(errorMessage: error, session: session)
             } else {
                 Policy.notifyUser(session: session, request: request)
             }
-        }
 
+        default:
```
Comment (agrinman) [inline Kryptonite/Silo.swift]: should not be default 👍

File: Kryptonite/Silo.swift [hunk 9]
```diff
@@ -190,17 +195,20 @@ class Silo {
...
         defer { log("response took \(Date().timeIntervalSince1970 - requestStart) seconds") }
-        var sign:SignResponse?
-        var me:MeResponse?
-        var gitSign:GitSignResponse?
         
-        if let signRequest = request.sign {
+        // the response type
+        var responseType:ResponseType
+        
+        // craft a response to the reuqest type
```
Comment (agrinman) [inline Kryptonite/Silo.swift]: 👍

File: Kryptonite/Silo.swift (hunk 1 above)
Comment (agrinman) [inline Kryptonite/Silo.swift]: should not be default 👍

File: Kryptonite/Silo.swift (hunk 2 above)
Comment (agrinman) [inline Kryptonite/Silo.swift]: it's not expected behavior as responseFor cannot create responses for noOp and unpair request bodies.

File: Kryptonite/Policy.swift [hunk 10]
```diff
@@ -229,13 +229,24 @@ class Policy {
...
             else {
                 return
             }
             
+            // ensure that session + request are auto-allowed by policy
+            switch pending.request.type {
+            case .ssh(let signRequest) where false == Policy.needsUserApproval(for: pending.session, and: signRequest):
+                break
+                
+            case .git where false == Policy.needsUserApproval(for: pending.session):
```
Comment (agrinman) [inline Kryptonite/Policy.swift]: yes 👍 , this decision should be made in one place, specifically in the policy class.

File: Notify/NotificationService.swift (hunk 3 above)
Comment (agrinman) [inline Notify/NotificationService.swift]: 👍

File: NotifyUI/NotificationViewController.swift (hunk 4 above)
Comment (agrinman) [inline NotifyUI/NotificationViewController.swift]: 👍  should not be default, but the NotificationContentUI extension will only be rendered for the autoApproval and approval notification categories and no others.

**Pull request merged**